src/java/htsjdk/samtools/cram/build/CompressionHeaderFactory.java

```diff
@@ -60,128 +50,121 @@
 
         final int readNameID = exCounter++;
         header.externalIds.add(readNameID);
         header.externalCompressors.put(readNameID, ExternalCompressor.createGZIP());
 
-        final int mateInfoID = exCounter++;
-        header.externalIds.add(mateInfoID);
-        header.externalCompressors.put(mateInfoID,
-                ExternalCompressor.createRANS(RANS.ORDER.ONE));
-
         header.encodingMap = new TreeMap<EncodingKey, EncodingParams>();
-        for (final EncodingKey key : EncodingKey.values())
+        for (final EncodingKey key : EncodingKey.values()) {
             header.encodingMap.put(key, NullEncoding.toParam());
+        }
 
         header.tMap = new TreeMap<Integer, EncodingParams>();
 
         { // bit flags encoding:
-            getOptimalIntegerEncoding(header, EncodingKey.BF_BitFlags, 0, records);
+            header.externalIds.add(exCounter);
+            header.externalCompressors.put(exCounter, ExternalCompressor.createRANS(RANS.ORDER.ONE));
+            header.encodingMap.put(EncodingKey.BF_BitFlags, ExternalIntegerEncoding.toParam(exCounter++));
         }
 
         { // compression bit flags encoding:
-            getOptimalIntegerEncoding(header, EncodingKey.CF_CompressionBitFlags, 0, records);
+            header.externalIds.add(exCounter);
+            header.externalCompressors.put(exCounter, ExternalCompressor.createRANS(RANS.ORDER.ONE));
+            header.encodingMap.put(EncodingKey.CF_CompressionBitFlags, ExternalIntegerEncoding.toParam(exCounter++));
         }
 
         { // ref id:
-
-            getOptimalIntegerEncoding(header, EncodingKey.RI_RefId, -2, records);
+            header.externalIds.add(exCounter);
+            header.externalCompressors.put(exCounter, ExternalCompressor.createRANS(RANS.ORDER.ZERO));
+            header.encodingMap.put(EncodingKey.RI_RefId, ExternalIntegerEncoding.toParam(exCounter++));
         }
 
         { // read length encoding:
-            getOptimalIntegerEncoding(header, EncodingKey.RL_ReadLength, 0, records);
+            header.externalIds.add(exCounter);
+            header.externalCompressors.put(exCounter, ExternalCompressor.createRANS(RANS.ORDER.ONE));
+            header.encodingMap.put(EncodingKey.RL_ReadLength, ExternalIntegerEncoding.toParam(exCounter++));
         }
 
         { // alignment offset:
-            if (sorted) { // alignment offset:
+            if (sorted) {
                 header.APDelta = true;
-                getOptimalIntegerEncoding(header, EncodingKey.AP_AlignmentPositionOffset, 0, records);
+
+                final int aStartID = exCounter++;
+                header.encodingMap.put(EncodingKey.AP_AlignmentPositionOffset,
+                        ExternalIntegerEncoding.toParam(aStartID));
+                header.externalIds.add(aStartID);
+                header.externalCompressors.put(aStartID, ExternalCompressor.createRANS(RANS.ORDER.ZERO));
```

Comment:
the two sides of this if else are basically identical, move the duplicated code out of the if statement.

looks like it should be replaced with
```
header.APDelta = sorted
if( sorted ) {
    log.debug("Assigned external id to alignment starts: " + aStartID);
}